[refactor]: update iroha_telemetry_derive to use syn2

VAmuzing opened the pull request:
## Description

Update with refactoring to remove proc_macro_error in favour of our `Emitter` wrapper

<!-- Just describe what you did. -->

<!-- Skip if the title of the PR is self-explanatory -->

### Linked issue

<!-- Duplicate the main issue and add additional issues closed by this PR. -->

Closes #3909 <!-- Replace with an actual number,  -->

<!-- Link if e.g. JIRA issue or  from another repository -->


### Checklist

- [x] I've read `CONTRIBUTING.md`
- [x] I've used the standard signed-off commit format (or will squash just before merging)
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

<!-- HINT:  Add more points to checklist for large draft PRs-->

<!-- USEFUL LINKS 
 - https://www.secondstate.io/articles/dco
 - https://discord.gg/hyperledger (please ask us any questions)
 - https://t.me/hyperledgeriroha (if you prefer telegram)
-->

VAmuzing commented:
Need to update failure tests as well if we proceed with these errors

DCNick3 commented on telemetry/derive/src/lib.rs:
```diff
@@ -1,15 +1,12 @@
-#[cfg(feature = "metric-instrumentation")]
-use proc_macro2::TokenStream as TokenStream2;
-use proc_macro_error::{abort, proc_macro_error};
+// use proc_macro::TokenStream;
```
Please remove commented code

VAmuzing replied:
Done

DCNick3 commented on telemetry/derive/src/lib.rs:
```diff
@@ -145,54 +142,91 @@ impl ToTokens for MetricSpec {
+                        emitter,
                         type_name,
-                        format!("Should return `Result`. Found {type_name}")
+                        "Should return `Result`. Found `{:?}`",
```
Shouldn't it be `{}`? At least it was the format used before..

VAmuzing replied:
Yeah, changed it

DCNick3 reviewed (commented):
Code looks good!

Though would like to look at the changed error messages (if any)

VAmuzing commented:
> Code looks good (aside from minor nitpicks)!
> 
> Though would like to look at the changed error messages (if any)

The errors contain additional lines as emitter does not abrupt macro compilation like `abort!` did.
This is the example:

> EXPECTED:
┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
error: Function must have at least one argument of type `WorldStateView`.
 --> tests/ui_fail/no_args.rs:4:1
  |
4 | fn execute() -> Result<(), ()> {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈

> ACTUAL OUTPUT:
> ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
> error: Function must have at least one argument of type `WorldStateView`.
>  --> tests/ui_fail/no_args.rs:4:1
>   |
> 4 | fn execute() -> Result<(), ()> {
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> error: At least one argument must be a `WorldStateView`.
>  --> tests/ui_fail/no_args.rs:3:1
>   |
> 3 | #[metrics(+"test_query", "another_test_query_without_timing")]
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   |
>   = note: this error originates in the attribute macro `metrics` (in Nightly builds, run with -Z macro-backtrace for more info)
> ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈

What do you think about printing additional errors?

DCNick3 commented:
This... doesn't seem right. Seems like the missing wsv argument is getting double-reported.

I guess you don't need this check, as it's also checked later: https://github.com/hyperledger/iroha/pull/4168/files#diff-6fe38d7a884371cd14b17c23142ac64383dee34a81b9f87a5b81e3eeebc25ff7R162-R168

VAmuzing commented:
> This... doesn't seem right. Seems like the missing wsv argument is getting double-reported.
> 
> I guess you don't need this check, as it's also checked later: https://github.com/hyperledger/iroha/pull/4168/files#diff-6fe38d7a884371cd14b17c23142ac64383dee34a81b9f87a5b81e3eeebc25ff7R162-R168

Adjusted `stderr` and added early return as the next one refers to the macro signature.
This was the error about empty signature is more clear.
